test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,485 @@
+    }
+  });
+
+  describe('User.login', function() {
+    it('Login a user from 1st user model by providing credentials', function(done) {
+      OneUser.login(commonCredentials)
+      .then(function(accessToken) {
+        assertGoodToken(accessToken, userFromOneModel);
+        done();
+      })
+      .catch(done);
+    });
+
+    it('Login a user from 2nd user model by providing credentials', function(done) {
+      AnotherUser.login(commonCredentials)
+      .then(function(accessToken) {
+        assertGoodToken(accessToken, userFromAnotherModel);
+        done();
+      })
+      .catch(done);
+    });
+
+    it('Login should only allow correct credentials', function(done) {
+      OneUser.login({email: 'foo@bar.com', password: 'invalid'})
+      .then(function(accessToken) {
+        expect(accessToken, 'accessToken').to.not.exist();
+        done();
+      })
+      .catch(function(err) {
+        expect(err, 'err').to.exist();
+        expect(err).to.have.property('code', 'LOGIN_FAILED');
+        done();
+      });
+    });
+  });
+
+  function assertGoodToken(accessToken, user) {
+    if (accessToken instanceof AccessToken) {
+      accessToken = accessToken.toJSON();
+    }
+    expect(accessToken).to.have.property('userId', user.id);
+    expect(accessToken).to.have.property('principalType', user.constructor.definition.name);
+    assert(accessToken.id);
+    assert.equal(accessToken.id.length, 64);
+  }
+
+  describe('User.logout', function() {
+    it('should logout a user from user model 1 without logging out user from model 2',
+    function(done) {
+      var userFromOneModelTokens, userFromAnotherModelTokens;
+      Promise.all([
+        OneUser.login(commonCredentials),
+        AnotherUser.login(commonCredentials),
+      ])
+      .spread(function(at1, at2) {
+        userFromOneModelTokens = at1;
+        userFromAnotherModelTokens = at2;
+        return OneUser.logout(userFromOneModelTokens.id);
+      })
+      .then(function() {
+        return Promise.all([
+          AccessToken.findById(userFromOneModelTokens.id),
+          AccessToken.findById(userFromAnotherModelTokens.id),
+        ]);
+      })
+      .spread(function(userFromOneModelTokens, userFromAnotherModelTokens) {
+        assert(!userFromOneModelTokens, 'accessToken should not exist after logging out user1');
+        assert(userFromAnotherModelTokens, 'accessToken should exist after logging out user1');
+        done();
+      })
+      .catch(done);
+    });
+  });
+
+  describe('Password Reset: ', function() {
+    describe('User.resetPassword(options)', function() {
+      var options = {
+        email: 'foo@bar.com',
+        redirect: 'http://foobar.com/reset-password',
+      };
+
+      it('should create a temp accessToken to allow a user to change password', function(done) {
+        OneUser.resetPassword({email: options.email})
+        .catch(function(err) {
+          assert(!err);
+        });
```

Comment:
Uff, this is rather ugly. I think we should use `Promise.all` to wait for two operations to finish:
 1. `OneUser.resetPassword` returns
 2. `resetPasswordRequest` is emitted

Thoughts?

```js
Promise.all([
  OneUser.resetPassword(...),
  waitForEvent(OneUser, 'resetPasswordRequest')
    .then(function(info) { return info.accessToken.user.getAsync(); })
    .then(/*...*/)
]);


// shared helper
function waitForEvent(emitter, name) {
  return new Promise(function(resolve, reject) {
    emitter.once(name, resolve);
  });
};
```